menhir-recover/recovery_new.ml

```diff
@@ -0,0 +1,196 @@
+  end
+
+  let depth = Lr1.tabulate (fun st ->
+    List.fold_left (fun pos (_, pos') -> max pos pos') 0 @@ Lr0.items (Lr1.lr0 st))
+
+  module State = struct
+    module S = Set.Make (
+      struct
+        let compare = Pervasives.compare
+        type t = Cost.t * Nonterminal.t
+      end)
+
+    type t = S.t ref * Trace.t option array
+
+    let empty () = (
+        ref S.empty, 
+        Array.make Nonterminal.count @@ Some {cost = Cost.infinite; items = []}
+      )
+
+    let is_empty (set, _) = S.is_empty !set
+
+    let get_min (set, arr) =
+      let (cost, nt) as elem = S.min_elt !set in
+      let i = Nonterminal.to_int nt in
+      let res = nt, Option.get arr.(i) in
+      arr.(i) <- None;
+      set := S.remove elem !set;
+      res
+
+    let add_elem (set, arr) (nt, tr) =
+      let i = Nonterminal.to_int nt in
+      match arr.(i) with
+        | Some tr' -> 
+          if tr.cost < tr'.cost then
+            (set := S.add (tr.cost, nt) @@ S.remove (tr'.cost, nt) !set;
+            arr.(i) <- Some tr)
+        | _ -> ()
+  end
+
+  let synthesize = 
+    Lr1.tabulate (fun st ->
+      List.fold_left (fun acc (prod, pos) ->
+        let item = st, prod, pos in
+        let cost, _ = S.solve (S.Tail(st, prod, pos)) in
+        (Production.lhs prod, {cost = cost; items = [item]}) :: acc
+      ) [] (Lr0.items (Lr1.lr0 st))
+    )
+
+  let step child parent =
+    let finish = List.filter_map (fun (prod, pos) ->
+        if Array.length (Production.rhs prod) <= pos then
+          None
+        else 
+          match (Production.rhs prod).(pos) with
+            | (N nt, _, _) -> 
+              let next = List.assoc (N nt) (Lr1.transitions parent) in
+              let cost, _ = S.solve (S.Tail(next, prod, pos + 1)) in
+              Some (nt, {cost = cost; items = [next, prod, pos + 1]})
+            | _ -> None
+      ) @@ Lr0.items (Lr1.lr0 parent) in
+    
+    let state = State.empty () in
+
+    List.iter (State.add_elem state) @@ synthesize child;
+
+    let rec aux acc =
+      if State.is_empty state then
+        acc
+      else 
+        let (nt, tr) = State.get_min state in
+        match List.find_opt (fun (nt', _) -> nt = nt') finish with
+          | Some (_, tr') -> 
+            aux (Trace.cat tr' tr :: acc)
+          | _ ->
+            (* TODO workaround for children with different positions in items*)
+            match List.assoc_opt (N nt) (Lr1.transitions parent) with 
+              | None -> aux acc 
+              | Some st' ->
+                synthesize st'
+                |> List.map (fun (nt, tr') -> nt, Trace.cat tr' tr)
+                |> List.iter @@ State.add_elem state;
+                aux acc
+    in
+    (List.fold_left Trace.min {cost = Cost.infinite; items = []} @@ aux []).items
+
+  let pred = Lr1.tabulate 
+    (fun st -> 
+      Lr1.fold (fun st' acc -> 
+        if List.fold_left (fun flag (_, st'') -> st = st'' || flag) false (Lr1.transitions st') then
```

Comment:
What do you think about the max length of lines? :thinking: 
Personally, I prefer 80, I accept 100 but it is even larger..

Reply:
Fixed